### PR TITLE
fix: disabled style miss when has href on Button

### LINKED ISSUE #39452: If you set the href attribute, the disabled style does not take effect
### Reproduction link

[Edit on CodeSandbox](https://codesandbox.io/s/antd-reproduction-template-forked-8ev4je?file=/index.js)

### Steps to reproduce

在button 的`disable={true}` 时，设置`href`，此时`disable`样式不生效

### What is expected?

Actually, I found this issue https://github.com/ant-design/ant-design/issues/35952 is similar. But in anted 5.0.5, the problem seems to have resurfaced and is not fully fixed.  What I expect is that when the button is not just a link type, setting disabled will take effect as well as the button type.

### What is actually happening?

button设置href后，diabled样式不生效

| Environment | Info |
| --- | --- |
| antd | 5.0.5 |
| React | 18.2.0 |
| System | Windows 11 22623.1028 |
| Browser | 107.0.1418.52 (正式版本) (64 位) |

### PR DESCRIPTION
**Pull request by BoyYangzai**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
 fix #39452 

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |       修复带有href属性的Button disabled时 style 不生效    |
| 🇨🇳 Chinese |     fix: disabled style miss when has href on Button      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by leo220yuyaodog** [comment]
@BoyYangzai I checked the pr. There's a little problem. The button has the disabled style, it is ok. But when the mouse hovers on the `a` tag text, it changes the style. It is better to keep the text and buttons consistent here.


image

**Comment by BoyYangzai** [comment]
I don’t seem to have the hover discoloration you mentioned here.
image
can you show your test code

**Comment by leo220yuyaodog** [comment]
> I don’t seem to have the hover discoloration you mentioned here. image can you show your test code

I use the [components/button/demo/disabled.tsx](https://github.com/ant-design/ant-design/blob/master/components/button/demo/disabled.tsx), 
Modify [line13-14](https://github.com/ant-design/ant-design/blob/555257113107821218899ffffa1acde11906f82e/components/button/demo/disabled.tsx#L12)
```javascript
<Space>
      <Button href="https://www.yuque.com/dashboard/notes">Default</Button>
      <Button href="https://www.yuque.com/dashboard/notes" disabled>Default(disabled)</Button>
</Space>
```

**Comment by leo220yuyaodog** [comment]
> > > I don’t seem to have the hover discoloration you mentioned here. image can you show your test code
> > 
> > 
> > I use the [components/button/demo/disabled.tsx](https://github.com/ant-design/ant-design/blob/master/components/button/demo/disabled.tsx), Modify [line13-14](https://github.com/ant-design/ant-design/blob/555257113107821218899ffffa1acde11906f82e/components/button/demo/disabled.tsx#L12)
> > ```js
> > <Space>
> >       <Button href="https://www.yuque.com/dashboard/notes">Default</Button>
> >       <Button href="https://www.yuque.com/dashboard/notes" disabled>Default(disabled)</Button>
> > </Space>
> > ```
> 
> I tried it and found that it is Space or when there is a parent element, it may be that the hover event caused by event capture is passed to the bottom layer, causing it to change color. I also added a disabled style to him after hover to solve it. Now it is succeeded，you can try again

Ok, now it is nice.

File: components/button/button.tsx [hunk 1]
```diff
@@ -254,12 +254,14 @@ const InternalButton: React.ForwardRefRenderFunction<unknown, ButtonProps> = (pr
...
 
+  const hrefAndDisabled = linkButtonRestProps.href !== undefined && mergedDisabled;
+
   const classes = classNames(
     prefixCls,
     hashId,
     {
       [`${prefixCls}-${shape}`]: shape !== 'default' && shape, // Note: Shape also has `default`
-      [`${prefixCls}-${type}`]: type,
+      [`${prefixCls}-${type}`]: !hrefAndDisabled && type,
```
Comment (MadCcc) [inline components/button/button.tsx:264]: Should remain type
```suggestion
      [`${prefixCls}-${type}`]: type,
```

**Comment by MadCcc** [comment]
Just follow this PR: https://github.com/ant-design/ant-design/pull/35975
The style is lost in V5

**Comment by BoyYangzai** [comment]
> Just follow this PR: #35975 . The style is lost in V5, and please provide a test case, thx~


i used styled-components &&& Increase priority of disabled :hover，eliminate the side effects of :not(:disabled) on disabled

Now，It should now work as expected 😄

**Comment by BoyYangzai** [comment]
@MadCcc 
image
now, ti's right on disabled

**Comment by MadCcc** [comment]
`&&&` is not a good pettern. Maybe `.ant-btn.ant-btn-disabled:hover` is better.
And how about hover pseudos, such as focus/active?

**Comment by BoyYangzai** [comment]
> `&&&` is not a good pettern. Maybe `.ant-btn.ant-btn-disabled:hover` is better. And how about hover pseudos, such as focus/active?

ok，that's good

**Comment by MadCcc** [comment]
And we need a debug demo to make sure this feature won't break in future versions.

File: components/button/style/index.tsx [hunk 2]
```diff
@@ -367,6 +370,14 @@ const genTextButtonStyle: GenerateStyle<ButtonToken, CSSObject> = (token) => ({
   },
 });
 
+// Href and Disabled
+const genDisabledButtonStyle: GenerateStyle<ButtonToken, CSSObject> = (token) => ({
+  ...genDisabledStyle(token),
+  '.ant-btn&:hover': {
```
[the remark this answers is not in the document]
Comment (BoyYangzai) [reply components/button/style/index.tsx]: > Is this priority enough？

这是可以的，已修改

**Comment by BoyYangzai** [comment]
@MadCcc 加上了demo，辛苦看下可以了不
image

**Comment by MadCcc** [comment]
Test still failed.

**Comment by BoyYangzai** [comment]
> Test still failed.

应该是Snapshot没有上传全，稍等，我马上更新

**Comment by BoyYangzai** [comment]
@MadCcc 快照更新了，使用的是npm run test:update，应该没有问题了 qwq

**Comment by BoyYangzai** [comment]
@MadCcc 哈咯哈咯 这个pr还需要review qwq

**Review by afc163 — APPROVED** [review]

**Pull request merged**